packages/neo-one-client/src/user/LocalUserAccountProvider.ts

```diff
@@ -823,6 +841,172 @@ export class LocalUserAccountProvider<TKeyStore extends KeyStore, TProvider exte
+    const outputArray = (): ReadonlyArray<OutputModel> => [];
+    const getMessageSize = (numInputs: number, numOutputs: number) =>
+      messageSize + numInputs * newInputSize + numOutputs * newOutputSize;
+
+    const updateOutput = (output: OutputModel, amount: BigNumber): OutputModel =>
+      new OutputModel({
+        asset: output.asset,
+        value: output.value.add(clientUtils.bigNumberToBN(amount, 8)),
+        address: output.address,
+      });
+
+    const consoladatableUnspentsIn = await this.provider.getUnspentOutputs(from.network, from.address, monitor);
+
+    const looseAssets = consoladatableUnspentsIn
+      .filter(
+        (unspent) =>
+          !inputs.some((input) => unspent.txid === input.txid && unspent.vout === input.vout) &&
+          !this.mutableUsedOutputs.has(`${unspent.txid}:${unspent.vout}`),
+      )
+      .reduce((acc, unspent) => {
+        if ((acc[unspent.asset] as ReadonlyArray<UnspentOutput> | undefined) !== undefined) {
+          return {
+            ...acc,
+            [unspent.asset]: acc[unspent.asset].concat([unspent]),
+          };
+        }
+
+        return {
+          ...acc,
+          [unspent.asset]: [unspent],
+        };
+      }, inputMap());
+
+    const { newInputs, updatedOutputs, looserAssets } = transactionUnsignedIn.outputs
+      .filter(
+        (output) =>
+          (looseAssets[common.uInt256ToString(output.asset)] as ReadonlyArray<UnspentOutput> | undefined) !==
+            undefined && common.uInt160ToString(output.address) === addressToScriptHash(from.address),
+      )
+      .reduce(
+        (acc, output) => {
+          const currentAsset = common.uInt256ToString(output.asset);
+          const toAdd = Math.min(
+            looseAssets[currentAsset].length,
+            Math.floor((byteLimit - getMessageSize(acc.newInputs.length, 0)) / newInputSize),
+          );
+          if (toAdd === 0) {
+            return acc;
+          }
+          const tempIns = acc.looserAssets[currentAsset]
+            .sort((coinA, coinB) => coinA.value.comparedTo(coinB.value))
+            .slice(0, toAdd);
+          const newUpdatedOutputs = acc.updatedOutputs.map((out) => {
+            if (out.address === output.address && out.asset === output.asset) {
+              return updateOutput(out, tempIns.reduce((left, right) => left.plus(right.value), new BigNumber('0')));
+            }
+
+            return out;
+          });
+          const evenLooserAssets = {
+            ...acc.looserAssets,
+            [currentAsset]: acc.looserAssets[currentAsset].slice(toAdd),
+          };
+
+          return {
+            newInputs: acc.newInputs.concat(this.convertInputs(tempIns)),
+            updatedOutputs: newUpdatedOutputs,
+            looserAssets: evenLooserAssets,
+          };
+        },
+        { newInputs: inputArray(), updatedOutputs: transactionUnsignedIn.outputs, looserAssets: looseAssets },
+      );
```

Comment:
Lots of comments above, roughly, here's how I'd do it:
```
const { newInputs, newOutputs, remainingAssetToUnspentOutputs } = transactionUnsignedIn.outputs.reduce(
  (acc, output) => {
    const asset = common.uInt256ToString(output.asset);
    const unspentOutputs = (
      (acc.remainingAssetToUnspentOutputs[asset] as ReadonlyArray<UnspentOutput> | undefined) === undefined ||
      common.uInt160ToString(output.address) !== addressToScriptHash(from.address)
    ) 
      ? []
      : acc.remainingAssetToUnspentOutputs[asset;
    const toAdd = Math.floor((byteLimit - getMessageSize({ numNewInputs: acc.newInputs.length })) / InputModel.size);
    const newInputs = unspentOutputs.slice(0, toAdd);

    return { 
      newInputs: acc.newInputs.concat(this.convertInputs(newInputs)),
      newOutputs: acc.newOutputs.concat([
        output.clone({ 
          value: output.value.plus(newInputs.reduce((acc, input) => acc.plus(input.value), new BigNumber('0'))),
        }),
      ]),
      remainingAssetToUnspentOutputs: {
        ...acc.remainingAssetToUnspentOutputs,
        [asset]: unspentOutputs.slice(toAdd),
      },
    };
  }, 
  { newInputs: [], newOutputs: [], remainingAssetToUnspentOutputs: assetToUnspentOutputs },
);
```

Reply:
Might need to add an explicit type to `reduce` so that typescript doesn't complain. (something like `reduce<{ newInputs: ReadonlyArray<Input>, ...}>`)